### PR TITLE
Explain Firefox how numeric input should work

### PR DESCRIPTION
**Pull request by Karakatiza666**
Is this a user-visible change (yes/no): yes

Fix https://github.com/feldera/feldera/issues/1522 : WebConsole: numeric inputs allow arbitrary text input in Firefox
Fix https://github.com/feldera/feldera/issues/1748 : WebConsole: Invariant failed setting non nullable value to null

### CODE REVIEW

**Comment by gz** [comment]
In this video I'm trying to clear the field by hitting backspace but it doesn't clear the last 4 digits in the field

[Screencast from 04-03-2024 10:04:52 AM.webm](https://github.com/feldera/feldera/assets/127654/42f0e09c-2146-4ca1-99ac-024b995d74fa)

**Comment by gz** [comment]
I'm trying to select the field and overwrite it with a single number but it doesn't let me
[Screencast from 04-03-2024 10:06:28 AM.webm](https://github.com/feldera/feldera/assets/127654/
e6143c00-5f21-4392-9972-7d336973c4b3)

**Comment by Karakatiza666** [comment]
The case for the first behavior is probably the minimum allowed value for that field

**Comment by Karakatiza666** [comment]
I need to update this behavior to enable starting to write a new number even though it doesn't fit in the allowed in the beginning

**Comment by mihaibudiu** [comment]
> The case for the first behavior is the minimum allowed value for that librdkafka option

Validation should be done at the end, not on every keystroke.
How can you type the number 2000?

**Comment by gz** [comment]
> The case for the first behavior is the minimum allowed value for that librdkafka option

I can't set it e.g., 1001 since it doesn't allow me to change the first digits inthe sequence

**Comment by gz** [comment]
It also feels more sluggish to me when I try this in the browser. 

How about we do the following: let users enter any values, including non-digit characters but highlight the form red when it's invalid and display an error code similar to how the other form elements do it (e.g., table input or forms for connectors)?

File: web-console/src/lib/components/streaming/import/SQLValueInput.tsx [hunk 1]
```diff
@@ -92,25 +83,19 @@ export const SQLValueInput = ({
...
         .with(SqlType.TINYINT, SqlType.SMALLINT, SqlType.INTEGER, () => ({
-          type: 'number',
-          ...props,
-          inputProps: {
-            ...props.inputProps,
+          ...numberInputProps({
+            ...props,
             ...(({ min, max }) => ({ min: min.toNumber(), max: max.toNumber() }))(numericRange(columnType)),
-            value: props.value === null ? '' : props.value, // Enable clearing of the input when setting value to null
+            value: props.value as null | number,
```
[the remark this answers is not in the document]
Comment (Karakatiza666) [reply web-console/src/lib/components/streaming/import/SQLValueInput.tsx:89]: A cast to a value of a type "number or null"

[the remark this answers is not in the document]
Comment (mihaibudiu) [reply web-console/src/lib/components/streaming/import/SQLValueInput.tsx:89]: then you should write (null | number)

[the remark this answers is not in the document]
Comment (mihaibudiu) [reply web-console/src/lib/components/streaming/import/SQLValueInput.tsx:89]: Is there an Option type? That would be much better.

File: web-console/src/lib/components/input/NumberInput.tsx [hunk 2]
```diff
@@ -0,0 +1,86 @@
+import { nonNull } from '$lib/functions/common/function'
+import { ChangeEventHandler, ClipboardEventHandler, KeyboardEventHandler } from 'react'
+import invariant from 'tiny-invariant'
+
+import { TextField, TextFieldProps } from '@mui/material'
+
+export const handleNumericKeyDown: KeyboardEventHandler<Element> = event => {
+  if (/Key[ACVXYZ]/.test(event.code) && event.ctrlKey) {
```
[the remark this answers is not in the document]
Comment (Karakatiza666) [reply web-console/src/lib/components/input/NumberInput.tsx:8]: Will add a comment. Shortcuts like Ctrl + C, Ctrl + V etc.

File: web-console/src/lib/components/streaming/import/SQLValueInput.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (Karakatiza666) [reply web-console/src/lib/components/streaming/import/SQLValueInput.tsx:89]: Not a native one. A user-space wrapper class could be used, but the native `null` works just as well - the typing is strong everywhere, so null-ability is explicit and can not be ignored

[the remark this answers is not in the document]
Comment (Karakatiza666) [reply web-console/src/lib/components/streaming/import/SQLValueInput.tsx:89]: The check for unguarded access to values that can be `null` and `undefined` is also enforced by typescript compiler, not just type checker

[the remark this answers is not in the document]
Comment (Karakatiza666) [reply web-console/src/lib/components/streaming/import/SQLValueInput.tsx:89]: `as (null | number)` gets reduced by the formatter to `as null | number`

**Comment by gz** [comment]
Thanks, looks like behavior works as intended now. I fear there is still some confusing behavior for a user:

- it seems to reset the back to a default value when switching views if the text field is currently red, this is surprising for a user if input just "corrects" itself

[Screencast from 04-03-2024 04:39:47 PM.webm](https://github.com/feldera/feldera/assets/127654/c1cf0950-45f8-428c-b295-a00b44fb317c)

- It looks like the UI knows about the min or max value of each field but it doesn't do a good job of communicating it back to the user (aside from making the text field red) e.g., how about we display what range is accepted as a small error text below the form field (like we do in other cases like in the screenshot)
[Screencast from 04-03-2024 04:45:47 PM.webm](https://github.com/feldera/feldera/assets/127654/2a36c2e8-033c-440a-9484-3b6447073376)
Screenshot from 2024-04-03 16-47-47


- Then even though an integer form field has errors, I am able to submit it (and it reverts it just deletes the option), I guess this is expected, but a little surprising that it lets me submit a form with errors (I'd have expected it to bring me back to that dialog as it does when for example you leave the bootstrap server field blank)
[Screencast from 04-03-2024 04:43:09 PM.webm](https://github.com/feldera/feldera/assets/127654/5ceb9e94-ef1b-4c7a-9c7d-f1ac6ff20fa5)

**Comment by gz** [comment]
There is some odd behavior, the first time a form is filled out the min/max range is not displayed but the text cell becomes red, after I submit the form the ranges are displayed

[Screencast from 04-05-2024 09:40:08 AM.webm](https://github.com/feldera/feldera/assets/127654/941c8584-1d68-4812-883e-b8ba3d9ee885)

**Comment by gz** [comment]
Given how many pitfalls there were with this PR: lets add test-cases for these cases before we merge (I don't see any currently). We should get into the habit to have tests for it otherwise long term stability and evolution of it will suffer.

**Comment by Karakatiza666** [comment]
> There is some odd behavior, the first time a form is filled out the min/max range is not displayed but the text cell becomes red, after I submit the form the ranges are displayed

This is intended - cell becoming red is the property of the input, error message with the range is the result of the form validation
I could make the input go orange rather than red If you feel the need for disambiguation?

**Comment by Karakatiza666** [comment]
Updated the PR. In some cases quickly typing up numbers gets laggy - that is due to implementation that depends on attempting to parse the input as a number on each keystroke. This is a tradeoff for a simpler and more robust implementation and can be optimized later if needed.

**Review by gz — APPROVED** [review]

**Pull request merged**